### DISCUSSION TITLE
Hierarchical notes

### DISCUSSION THREAD

**Discussion by khramoff** [foambubble/foam#1034] [General]
I'm using notes for my studies and organizing my thoughts. 
My use case is not unique, obviously. But my approach is somewhat different from the norm.

I extensively use TIPS (TRIZ, here is a wiki [article](https://en.wikipedia.org/wiki/TRIZ) about it) approach externalized onto the computer science. It calls for the certain hierarchy of thought I have trouble organizing minimally, with little overhead coming from the tool. I'm yet to find the best one, but for now - Foam ties into it better than others. 

I stumbled upon Dendron while looking for the tools that will suit my needs.  
Dendron is somewhat monstrous in its features and tries to be all the things at the same time.

But one feature of the dendron has caught my attention. 
**A hierarchical note-taking.** 
Basically, it was a starting point for dendron.

The link to the original blog post with all the rationale by Kevin Slin - [here](https://www.kevinslin.com/notes/3dd58f62-fee5-4f93-b9f1-b0f0f59a9b64.html) 

To expand on the topic further. Hierarchical note-taking assumes the existence of schemes for notes. 
Schemas are the skeletons for similar topics and are extremely useful for organizing knowledge into the set dimensionality of concepts. Notes can be organized similarly across directories/collections of notes. 

I would like to see this in Foam. 
I realize perfectly well, that Foam does not try to dictate to the user what to do with their notes. Or how to approach note-taking. 

The main point of this feature request is the call for discussion. 
It appears to me that the hierarchy first approach is a core feature of a note-taking tool. The one aspect that concerns me is the philosophy behind Foam - will it tie into it? 

Foam is a very minimalistic tool, and it does not come in my way. I tell my note-taking system what to do, I organize it as I like. 
I would love this feature to be a part of Foam. 

Maybe, this approach can be abstracted and implemented as a separate **additional** component (extension). 

I'm unable to achieve this with features that are available for me, and I made multiple tries to project hierarchical note-taking onto existing features. 

To summarize, I would love to hear other thoughts regarding this topic.

**Comment by maciejmatczak** [comment]
# Hierarchy

I would truly like a directories to be a first class citizen in building notes hierarchy.

VS Code does make a good use of it - `CTRL+P` and start typing using your folder names (hierarchy node names) and name of the file and you quickly jump to a note.

And I do use it for my notes. For example as `dev/python/debugging.md`.

I also additionally experiment in graph visibility. As folders are not yet (?) reflected as proper nodes in graph, I additionally keep an indexing files:

 - `root.md` at root
 - `dev/dev.md` that references `root.md`
 - `dev/python/python.md` that references `dev/dev.md`
 - `dev/python/debugging.md` that references `dev/dev.md`

That does look good (in graph). But is for sure maintenance burden that could be avoided.

# Schemas

I do also use schemas/templates. Currently I created a set of VS Code tasks and mini python scripts. My actions (tied with schemas and with directories) are as follows:

 - create weekly journal, by copying everything from previous, updating a date (I would like to cleanup some particular sections, but currently settled down at manual selection and deletion)
 - create new project - new .md file in `projects` folder based on `templates/project.md.j2` template (yup, using a Python Jinja2 for dynamic parts in templates/schemas)
 - create new area - similar as above, generally a pseudo PARA; special template for that
 - create note - generic note template, the one that I am using in _second brain_ part of my notes, `knowledge` folder + additional hierarchy
 - create meeting - creates a meeting notes in `meetings` folder, also different template, date based naming convention + adds a backlink under currently opened file and current line

I can say - my use case is a bit complicated. Predetermined actions tied with templates and particular folders, sometimes additional information needed. But from what I can see, is that in every case a **template is correlated with a folder where new note is created**!

It would really make sense for me, if Foam could use a folder based templates, the first template existing in the folders hierarchy, starting from the place where a new note is about to be created up to the root.

If such template would be flexible enough, not only to use some templating language, special variables, but also could take a part in generating a "creation form", that would be truly magnificent (going crazy on that one ;) ).

By "creation form" I mean an additional inputs provided for the users when they would like to create a new note from Foam's _new note_ action. Currently for those I use a VS Code's tasks inputs, which gives me quite big flexibility, yet I hit limitations (like no folder picker available yet: https://github.com/microsoft/vscode/issues/71461).
